mobilitypath_publisher/src/mobilitypath_publisher.cpp

```diff
@@ -0,0 +1,138 @@
+        cav_msgs::MobilityPath mobility_path = mobilityPathMessageGenerator(latest_trajectory);
+        mob_path_pub_.publish(mobility_path);
+    }
+
+    void MobilityPathPublication::currentpose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+    {
+        current_pose_ = msg;
+    }
+
+    cav_msgs::MobilityPath MobilityPathPublication::mobilityPathMessageGenerator(cav_msgs::TrajectoryPlan trajectory_plan)
+    {
+        cav_msgs::MobilityPath mobility_path_msg;
+        mobility_path_msg.header = composeMobilityHeader();
+        
+
+        std::vector<cav_msgs::TrajectoryPlanPoint> traj_points;
+        traj_points = trajectory_plan.trajectory_points;
+        cav_msgs::Trajectory mob_path_traj = TrajectoryPlantoTrajectory(traj_points);
+        mobility_path_msg.trajectory = mob_path_traj;
+
+        return mobility_path_msg;
+    }
+
+    cav_msgs::MobilityHeader MobilityPathPublication::composeMobilityHeader(){
+        cav_msgs::MobilityHeader header;
+        header.sender_id = sender_id;
+        header.recipient_id = recipient_id;
+        header.sender_bsm_id = sender_bsm_id;
+        // random GUID that identifies this particular plan for future reference
+        header.plan_id = boost::uuids::to_string(boost::uuids::random_generator()());
+        header.timestamp = ros::Time::now().toNSec()/1000; //time in millisecond
+        
+        return header;
+    }
+
+    cav_msgs::Trajectory MobilityPathPublication::TrajectoryPlantoTrajectory(std::vector<cav_msgs::TrajectoryPlanPoint> traj_points){
+        cav_msgs::Trajectory traj;
+        cav_msgs::LocationECEF ecef_location = TrajectoryPointtoECEF(traj_points[0]);
+
+        if (traj_points.size()<2){
+            ROS_WARN("Received Trajectory Plan is too small");
+            traj.offsets = {};
+        }
+        else{
+            for (int i=1; i<traj_points.size(); i++){
+                if (trajectoryPoseCheck(traj_points[i])){
+                    cav_msgs::LocationOffsetECEF offset;
+                    cav_msgs::LocationECEF new_point = TrajectoryPointtoECEF(traj_points[i]);
+                    offset.offset_x = new_point.ecef_x - ecef_location.ecef_x;
+                    offset.offset_y = new_point.ecef_y - ecef_location.ecef_y;
+                    offset.offset_z = new_point.ecef_z - ecef_location.ecef_z;
+                    traj.offsets.push_back(offset);
+                }
+                else{
+                    ecef_location = TrajectoryPointtoECEF(traj_points[i]);
+                    traj.offsets = {};
+                }
+            }
+        }
+        
+        
+        traj.location = ecef_location;
+        return traj;
+    }
+
+    bool MobilityPathPublication::trajectoryPoseCheck(cav_msgs::TrajectoryPlanPoint point){
+        if (fabs(current_pose_->pose.position.x - point.x) > x_threshold || fabs(current_pose_->pose.position.y - point.y) > y_threshold)
+        {
+            return false;
+        }
+        return true;
```

Comment:
I'm not sure the check in this function makes sense. You are trying to evaluate which points to include in the path, but the x,y positions are in the map frame and could be any value. So you could reject large chunks of the path based on the orientation of the vehicle in the map. 

I think the main goal of this function is to only include points which are still to be executed. So it would make more sense to have a function that tells you which is the nearest point in the trajectory in both space and time. Select that as the current point, then keep all future points in the path but drop the previous points.

Reply:
After reviewing the design doc, my initial comment above isn't entirely accurate but there is still a fundamental issue.
The function "trajectoryPoseCheck" doesn't actually perform the way described in the highlevel design doc which used downtrack/crosstrack instead of x,y,z to perform the comparison.
What you are doing is comparing vehicle position in the map frame, to trajectory points in the map frame. This is less useful because all your are evaluating is the distance from each individual point rather than the distance from the curve you wish to follow or your error in time. Additionally, the uncertainty values in your config file will not map correctly as the map frame X-axis is rarely going to be aligned with your roadway.
I would recommend that you use the trackPos function in carma_wm/Geometry.h to compute the downtrack/crosstrack position of your vehicle relative to the entire path. Then have a crosstrack threshold, time threshold, and downtrack threshold that can be used to evaluate the vehicle's performance against the planned trajectory

Reply:
removed the check